planner/core/exhaust_physical_plans.go

```diff
@@ -262,18 +262,26 @@ func (p *LogicalJoin) getHashJoins(prop *property.PhysicalProperty) []PhysicalPl
 	}
 	joins := make([]PhysicalPlan, 0, 2)
 	switch p.JoinType {
-	case SemiJoin, AntiSemiJoin, LeftOuterSemiJoin, AntiLeftOuterSemiJoin, LeftOuterJoin:
-		joins = append(joins, p.getHashJoin(prop, 1))
+	case SemiJoin, AntiSemiJoin, LeftOuterSemiJoin, AntiLeftOuterSemiJoin:
+		joins = append(joins, p.getHashJoin(prop, 1, false))
+	case LeftOuterJoin:
+		joins = append(joins, p.getHashJoin(prop, 1, false))
+		if len(p.EqualConditions) > 0 {
+			joins = append(joins, p.getHashJoin(prop, 1, true))
+		}
 	case RightOuterJoin:
-		joins = append(joins, p.getHashJoin(prop, 0))
+		joins = append(joins, p.getHashJoin(prop, 0, false))
+		if len(p.EqualConditions) > 0 {
```

Comment:
I do not understand why we can only use the outer to build the hashtable when `len(p.EqualConditions) > 0` is true.

Reply:
```
// ExplainInfo implements Plan interface.
func (p *PhysicalHashJoin) ExplainInfo() string {
	buffer := new(bytes.Buffer)

	if len(p.EqualConditions) == 0 {
		buffer.WriteString("CARTESIAN ")
	}
```
This code means the hash join would be converted to `CARTESIAN` if `len(p.EqualConditions) == 0 `. So we get rid of this case when taking the outer hash join.

Reply:
IMHO, OuterJoin will not be a `CARTESIAN`?

Reply:
Maybe, but it does not take the hash join.

Reply:
> but it does not take the hash join

I can't get your point.

Reply:
> > but it does not take the hash join
> 
> I can't get your point.

in that case (`len(p.EqualConditions) == 0 `), it will take the `CARTESIAN` join, instead of the hash join.

Reply:
Do you mean that if `LogicalJoin.cartesianJoin` is true, the physical join will not be a HashJoin?

Reply:
I check the code, the code here cannot generate `CARTESIAN join`, so I delete the judgement (`len(p.EqualConditions) > 0`).